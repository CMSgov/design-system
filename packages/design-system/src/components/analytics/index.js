export { default as sendEvent } from './Analytics';
