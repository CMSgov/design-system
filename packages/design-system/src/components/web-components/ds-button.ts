import { define } from 'preactement';
import Button from '../Button/Button';

const attributes = [
  'class-name',
  'disabled',
  'href',
  'is-alternate',
  'on-dark',
  'size',
  'type',
  'variation',
  'analytics',
  'analytics-label-override',
  'analytics-event-type-override',
  'analytics-parent-heading',
  'analytics-parent-type',
];

define('ds-button', () => Button, { attributes });
