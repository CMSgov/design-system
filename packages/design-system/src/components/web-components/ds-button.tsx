import React from 'react';
import { define } from './preactement/define';
import Button, { ButtonProps } from '../Button/Button';

const attributes = [
  'class-name',
  'disabled',
  'href',
  'is-alternate',
  'is-on-dark',
  'size',
  'type',
  'variation',
  'target',
  'analytics',
  'analytics-label-override',
  'analytics-event-type-override',
  'analytics-parent-heading',
  'analytics-parent-type',
];

// Mapping `onDark` to `isOnDark` because props starting with "on" indicate an event handler and tests fail due to this expectation
interface WrapperProps extends Omit<ButtonProps, 'isAlternate' | 'onDark' | 'analytics'> {
  analytics?: string;
  isAlternate?: string;
  isOnDark?: string;
}

const Wrapper = ({ isAlternate, isOnDark, analytics, ...otherProps }: WrapperProps) => (
  <Button
    {...otherProps}
    {...{
      isAlternate: isAlternate && Boolean(JSON.parse(isAlternate)),
      onDark: isOnDark && Boolean(JSON.parse(isOnDark)),
      analytics: analytics && Boolean(JSON.parse(analytics)),
    }}
  />
);

define('ds-button', () => Wrapper, { attributes });
