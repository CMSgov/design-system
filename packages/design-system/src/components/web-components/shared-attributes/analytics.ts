export const analyticsAttrs = [
  'analytics',
  'analytics-label-override',
  'analytics-event-type-override',
  'analytics-parent-heading',
  'analytics-parent-type',
];
