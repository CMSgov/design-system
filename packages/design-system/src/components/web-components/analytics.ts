import { AnalyticsEvent } from '../analytics';
import { config } from '../config';

/**
 * The web component event configuration to pass into the define function's `events`
 * option that will allow users to override the default analytics function by creating
 * an event listener for the `ds-analytics-event` event and calling the `preventDefault`
 * function.
 *
 * For example,
 *
 * define('ds-button', () => Wrapper, { attributes, events: ['onClick', onAnalyticsEvent] });
 */
export const onAnalyticsEvent = [
  'onAnalyticsEvent',
  (event: AnalyticsEvent) => {
    let callDefaultAnalyticsFunction = true;

    function preventDefault() {
      callDefaultAnalyticsFunction = false;
    }

    // Give event listeners a chance to prevent default before we execute the default
    // analytics function from the config.
    setTimeout(() => {
      if (callDefaultAnalyticsFunction) {
        config().defaultAnalyticsFunction(event);
      }
    }, 10);

    return {
      detail: { event },
      preventDefault,
    };
  },
] as const;
