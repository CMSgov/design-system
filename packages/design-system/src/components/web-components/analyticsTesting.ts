import { UtagContainer } from '../analytics/index';
import { config } from '../config';

export function testAnalyticsEvents({ tagName, renderComponent, configField, triggerAction }) {
  function before() {
    config({ [configField]: true });
    const tealiumMock = jest.fn();
    (window as any as UtagContainer).utag = {
      link: tealiumMock,
    };
    jest.useFakeTimers();
    return tealiumMock;
  }

  function after() {
    config({ [configField]: false });
    jest.resetAllMocks();
    jest.useRealTimers();
  }

  function waitForAnalytics() {
    jest.advanceTimersToNextTimer();
    new Promise((resolve) => setTimeout(resolve, 0));
  }

  it('sends analytics event', async () => {
    const tealiumMock = before();
    renderComponent();
    await triggerAction();
    await waitForAnalytics();
    expect(tealiumMock.mock.calls[0]).toMatchSnapshot();
    after();
  });

  it('disables analytics event tracking', async () => {
    const tealiumMock = before();
    renderComponent({ analytics: 'false' });
    await triggerAction();
    await waitForAnalytics();
    expect(tealiumMock).not.toHaveBeenCalled();
    after();
  });

  it('setting analytics to true overrides flag value', async () => {
    const tealiumMock = before();
    config({ [configField]: false });
    renderComponent({ analytics: 'true' });
    await triggerAction();
    await waitForAnalytics();
    expect(tealiumMock).toHaveBeenCalled();
    after();
  });

  it('overrides analytics event tracking on open', async () => {
    const tealiumMock = before();
    renderComponent({ 'analytics-label-override': 'alternate content' });
    await triggerAction();
    await waitForAnalytics();
    expect(tealiumMock.mock.calls[0]).toMatchSnapshot();
    after();
  });

  it.skip('allows default analytics function to be ovewridden', async () => {
    const tealiumMock = before();
    let analyticsEvent;
    renderComponent();
    document.querySelector(tagName).addEventListener('ds-analytics-event', (event: any) => {
      event.preventDefault();
      analyticsEvent = event.detail.event;
    });
    await triggerAction();
    await waitForAnalytics();
    expect(tealiumMock).not.toHaveBeenCalled();
    expect(analyticsEvent).toBeDefined();
    expect(analyticsEvent).toMatchSnapshot();
    after();
  });
}
