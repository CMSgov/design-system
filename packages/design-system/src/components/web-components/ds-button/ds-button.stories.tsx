import { useEffect } from 'react';
import WebComponentDocTemplate from '../../../../../../.storybook/docs/WebComponentDocTemplate.mdx';
import { action } from '@storybook/addon-actions';
import './ds-button';
import { webComponentDecorator } from '../storybook';
import {
  analyticsEventDocs,
  analyticsOverrideArgTypes,
  analyticsParentDataArgTypes,
} from '../shared-attributes/storybook';

export default {
  title: 'Web Components/ds-button',
  argTypes: {
    'class-name': {
      description: 'Additional classes to be added to the root button element.',
      control: 'text',
    },
    'is-alternate': {
      description:
        'Applies the alternate color to a Button. By default, Button uses the `main` color.',
      control: 'boolean',
    },
    size: {
      description: 'A string corresponding to Button size classes.',
      options: [undefined, 'big', 'small'],
      control: { type: 'radio' },
    },
    variation: {
      description: 'A string corresponding to Button variation classes.',
      options: [undefined, 'solid', 'ghost'],
      control: { type: 'radio' },
    },
    ...analyticsOverrideArgTypes,
    ...analyticsParentDataArgTypes,
  },
  args: {},
  parameters: {
    docs: {
      page: WebComponentDocTemplate,
      description: {
        component: `For information about how and when to use this component, [refer to its full documentation page](https://design.cms.gov/components/button/).`,
      },
      componentEvents: {
        ...analyticsEventDocs,
      },
    },
  },
  decorators: [webComponentDecorator],
};

const Template = (args) => {
  useEffect(() => {
    const onClick = (event) => {
      action('ds-click')(event);
    };
    const button = document.querySelector('ds-button');
    button.addEventListener('ds-click', onClick);
    return () => {
      button.removeEventListener('ds-click', onClick);
    };
  });
  return <ds-button {...args}>{args.children ?? <>Your button text is here</>}</ds-button>;
};

export const Default = Template.bind({});
