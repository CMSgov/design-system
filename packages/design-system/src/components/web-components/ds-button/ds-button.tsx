import { define } from '../preactement/define';
import Button, { ButtonProps } from '../../Button/Button';
import { analyticsOverrideAttrs, analyticsParentDataAttrs } from '../shared-attributes/analytics';
import { onAnalyticsEvent } from '../analytics';

const attributes = [
  'class-name',
  'disabled',
  'href',
  'is-alternate',
  'is-on-dark',
  'size',
  'type',
  'variation',
  'target',
  ...analyticsOverrideAttrs,
  ...analyticsParentDataAttrs,
];

// Mapping `onDark` to `isOnDark` because props starting with "on" indicate an event handler and tests fail due to this expectation
interface WrapperProps extends Omit<ButtonProps, 'isAlternate' | 'onDark' | 'analytics'> {
  analytics?: string;
  isAlternate?: string;
  isOnDark?: string;
}

const Wrapper = ({ isAlternate, isOnDark, analytics, ...otherProps }: WrapperProps) => (
  <Button
    {...otherProps}
    {...{
      isAlternate: isAlternate && Boolean(JSON.parse(isAlternate)),
      onDark: isOnDark && Boolean(JSON.parse(isOnDark)),
      analytics: analytics && Boolean(JSON.parse(analytics)),
    }}
  />
);

define('ds-button', () => Wrapper, { attributes, events: ['onClick', onAnalyticsEvent] });
