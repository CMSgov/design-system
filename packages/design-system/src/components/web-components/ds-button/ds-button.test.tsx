import { UtagContainer } from '../../analytics/index';
import { config } from '../../config';
import { fireEvent, render, screen } from '@testing-library/react';
import './ds-button';
import { testAnalytics } from '../analyticsTesting';

/* eslint-disable @typescript-eslint/no-namespace */
declare global {
  namespace JSX {
    interface IntrinsicElements {
      'ds-button': any;
    }
  }
}
/* eslint-enable */

const defaultProps = {
  children: 'Foo',
};

function renderButton(props = {}) {
  return render(<ds-button {...defaultProps} {...props} />);
}

describe('Button', () => {
  it('renders as button', () => {
    const { asFragment } = renderButton();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders as submit button', () => {
    renderButton({ type: 'submit' });
    expect(screen.getByRole('button').getAttribute('type')).toEqual('submit');
  });

  it('renders disabled button', () => {
    const { asFragment } = renderButton({ disabled: true });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders as an anchor with custom prop', () => {
    const { asFragment } = renderButton({
      href: '/example',
      target: '_blank',
      type: 'submit',
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders disabled anchor correctly', () => {
    const { asFragment } = renderButton({
      href: '#!',
      disabled: true,
      children: 'Link button',
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('applies additional classes', () => {
    renderButton({ 'class-name': 'foobar' });
    const button = screen.getByRole('button');
    expect(button.classList.contains('foobar')).toBe(true);
  });

  it('applies variation classes', () => {
    renderButton({ variation: 'solid' });
    const button = screen.getByRole('button');
    expect(button.classList.contains('ds-c-button--solid')).toBe(true);
  });

  it('applies size classes', () => {
    renderButton({ size: 'small' });
    const button = screen.getByRole('button');
    expect(button.classList.contains('ds-c-button--small')).toBe(true);
  });

  it('applies disabled, inverse, alternate, and variation classes together', () => {
    renderButton({
      href: '#!',
      disabled: true,
      'is-on-dark': true,
      'is-alternate': true,
      variation: 'ghost',
    });
    const link = screen.getByRole('link');
    expect(link.hasAttribute('href')).toBe(false);
    expect(link.classList.contains('ds-c-button--ghost')).toBe(true);
    expect(link.classList.contains('ds-c-button--on-dark')).toBe(true);
    expect(link.classList.contains('ds-c-button--alternate')).toBe(true);
    expect(link.classList.contains('ds-c-button')).toBe(true);
  });

  it('fires a custom click event on click', () => {
    renderButton();
    const buttonRoot = document.querySelector('ds-button');
    const buttonEl = screen.getByRole('button');
    const mockHandler = jest.fn();
    buttonRoot.addEventListener('ds-click', mockHandler);
    fireEvent.click(buttonEl);
    expect(mockHandler).toHaveBeenCalledTimes(1);
    buttonRoot.removeEventListener('ds-click', mockHandler);
  });

  it('fires a custom analytics event on click', () => {
    renderButton({ analytics: 'true' });
    const buttonRoot = document.querySelector('ds-button');
    const buttonEl = screen.getByRole('button');
    const mockHandler = jest.fn();
    buttonRoot.addEventListener('ds-analytics-event', mockHandler);
    fireEvent.click(buttonEl);
    expect(mockHandler).toHaveBeenCalledTimes(1);
    buttonRoot.removeEventListener('ds-analytics-event', mockHandler);
  });

  describe('Analytics', () => {
    beforeEach(() => {
      config({ buttonSendsAnalytics: true });
    });

    afterEach(() => {
      config({ buttonSendsAnalytics: false });
    });

    testAnalytics('sends button analytics event', async ({ tealiumMock, waitForAnalytics }) => {
      renderButton();
      fireEvent.click(screen.getByRole('button'));
      await waitForAnalytics();
      expect(tealiumMock.mock.calls[0]).toMatchSnapshot();
    });

    testAnalytics('sends anchor analytics event', async ({ tealiumMock, waitForAnalytics }) => {
      renderButton({ href: '#/somewhere-over-the-rainbow' });
      fireEvent.click(screen.getByRole('link'));
      await waitForAnalytics();
      expect(tealiumMock.mock.calls[0]).toMatchSnapshot();
    });

    testAnalytics(
      'disables analytics event tracking',
      async ({ tealiumMock, waitForAnalytics }) => {
        renderButton({ analytics: 'false' });
        fireEvent.click(screen.getByRole('button'));
        await waitForAnalytics();
        expect(tealiumMock).not.toHaveBeenCalled();
      }
    );

    testAnalytics(
      'setting analytics to true overrides flag value',
      async ({ tealiumMock, waitForAnalytics }) => {
        config({ buttonSendsAnalytics: false });
        renderButton({ analytics: 'true' });
        fireEvent.click(screen.getByRole('button'));
        await waitForAnalytics();
        expect(tealiumMock).toHaveBeenCalled();
      }
    );

    testAnalytics(
      'overrides analytics event tracking on open',
      async ({ tealiumMock, waitForAnalytics }) => {
        renderButton({ 'analytics-label-override': 'alternate content' });
        fireEvent.click(screen.getByRole('button'));
        await waitForAnalytics();
        expect(tealiumMock.mock.calls[0]).toMatchSnapshot();
      }
    );

    testAnalytics(
      'allows default analytics function to be ovewridden',
      async ({ tealiumMock, waitForAnalytics }) => {
        let analyticsEvent;
        renderButton();
        document.querySelector('ds-button').addEventListener('ds-analytics-event', (event: any) => {
          event.preventDefault();
          analyticsEvent = event.detail.event;
        });
        fireEvent.click(screen.getByRole('button'));
        await waitForAnalytics();
        expect(tealiumMock).not.toHaveBeenCalled();
        expect(analyticsEvent).toMatchSnapshot();
      }
    );

    testAnalytics(
      'passes along parent heading and type',
      async ({ tealiumMock, waitForAnalytics }) => {
        const analyticsParentHeading = 'Hello World';
        const analyticsParentType = 'div';
        renderButton({
          'analytics-parent-heading': analyticsParentHeading,
          'analytics-parent-type': analyticsParentType,
        });
        fireEvent.click(screen.getByRole('button'));
        await waitForAnalytics();
        expect(tealiumMock).toBeCalledWith(
          expect.objectContaining({
            parent_component_heading: analyticsParentHeading,
            parent_component_type: analyticsParentType,
          })
        );
      }
    );
  });
});
