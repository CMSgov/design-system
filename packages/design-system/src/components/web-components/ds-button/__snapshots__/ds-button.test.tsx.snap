// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Analytics allows default analytics function to be ovewridden 1`] = `
Object {
  "button_style": "default",
  "button_type": "button",
  "event_extension": "Design system integration",
  "event_name": "button_engagement",
  "parent_component_heading": " ",
  "parent_component_type": " ",
  "text": "Foo",
}
`;

exports[`Button Analytics overrides analytics event tracking on open 1`] = `
Array [
  Object {
    "button_style": "default",
    "button_type": "button",
    "event_extension": "Design system integration",
    "event_name": "button_engagement",
    "parent_component_heading": " ",
    "parent_component_type": " ",
    "text": "alternate content",
  },
]
`;

exports[`Button Analytics sends anchor analytics event 1`] = `
Array [
  Object {
    "button_style": "default",
    "button_type": "link",
    "event_extension": "Design system integration",
    "event_name": "button_engagement",
    "link_url": "#/somewhere-over-the-rainbow",
    "parent_component_heading": " ",
    "parent_component_type": " ",
    "text": "Foo",
  },
]
`;

exports[`Button Analytics sends button analytics event 1`] = `
Array [
  Object {
    "button_style": "default",
    "button_type": "button",
    "event_extension": "Design system integration",
    "event_name": "button_engagement",
    "parent_component_heading": " ",
    "parent_component_type": " ",
    "text": "Foo",
  },
]
`;

exports[`Button renders as an anchor with custom prop 1`] = `
<a
  class="ds-c-button"
  href="/example"
  target="_blank"
>
  <slot />
</a>
`;

exports[`Button renders as button 1`] = `
<button
  class="ds-c-button"
  type="button"
>
  <slot />
</button>
`;

exports[`Button renders disabled anchor correctly 1`] = `
<a
  aria-disabled="true"
  class="ds-c-button"
  role="link"
>
  <slot />
</a>
`;

exports[`Button renders disabled button 1`] = `
<button
  class="ds-c-button"
  disabled=""
  type="button"
>
  <slot />
</button>
`;
